pkg/fileutils/filelock.go

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fileutils
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+// FileLock defines a file lock implemented by syscall.Flock
+type FileLock struct {
+	fileName string
+	fd       *os.File
+}
+
+// NewFileLock create a FileLock instance
+func NewFileLock(name string) *FileLock {
+	return &FileLock{
+		fileName: name,
+	}
+}
+
+// Lock locks file.
+// If the file is already locked, the calling goroutine blocks until the file is unlocked.
+// If lock has been invoked without unlock, lock again will return an error.
+func (l *FileLock) Lock() error {
+	var (
+		fd  *os.File
+		err error
+	)
+	if fd, err = os.Open(l.fileName); err != nil {
+		return err
+	}
+
+	if l.fd != nil {
```

Comment:
Why not put it ahead and you should not have to open the file and close it immediately.